onnxscript/rewriter/fuse_pad_into_conv.py

```diff
@@ -0,0 +1,296 @@
+"""
+
+from __future__ import annotations
+
+from typing import List, Sequence
+
+import numpy as np
+import onnx_ir as ir
+
+from onnxscript.rewriter import pattern as orp
+
+
+def fill_pads_with_axes(pads: Sequence[int], axes: Sequence[int], rank: int) -> List[int]:
+    new_pads = []
+    for axis in range(rank):
+        if axis not in axes:
+            start_value = end_value = 0
+        else:
+            start_value = pads[axes.index(axis)]
+            end_value = pads[axes.index(axis) + len(axes)]
+        pad_len = len(new_pads) // 2
+        new_pads.insert(pad_len + axis, end_value)
+        new_pads.insert(axis, start_value)
+    return new_pads
+
+
+def read_conv_attributes(ir_conv: ir.Node) -> dict[str, Sequence[int] | str]:
+    # Read attributes
+    attributes = {}
+    ir_attributes = ir_conv.attributes
+    attributes["kernel_shape"] = ir_attributes.get_ints(
+        "kernel_shape", ir_conv.inputs[1].shape[2:]
+    )
+    attributes["strides"] = ir_attributes.get_ints(
+        "strides", [1] * len(ir_conv.inputs[0].shape[2:])
+    )
+    attributes["auto_pad"] = ir_attributes.get_string("auto_pad", "NOTSET")
+    if "pads" in ir_attributes:
+        attributes["pads"] = ir_attributes.get_ints("pads")
+    return attributes
+
+
+class _FusePadConvBase(orp.RewriteRuleClassBase):
+    """Interface for PadConv nodes fusion."""
+
+    def __init__(self, as_function: bool = False):
+        # Remove nodes is set to False to remove unused nodes after the rewrite.
+        super().__init__(remove_nodes=False, as_function=as_function)
+
+    def rewrite(
+        self, op: ir.tape.Tape, x: ir.Value, pad: ir.Value, conv: ir.Value
+    ) -> ir.Value:
+        pad_node = pad.producer()
+        conv_node = conv.producer()
+
+        # Retrieve the padding and axes
+        x_rank = len(x.shape)
+        pad_pads = pad_node.inputs[1].const_value.numpy().tolist()
+        if len(pad_node.inputs) > 3 and (axes := pad_node.inputs[3]) is not None:
+            axes = [x if x >= 0 else x_rank + x for x in axes.const_value.numpy()]
+        else:
+            axes = list(range(x_rank))
+
+        # Fulfill pad_pads in every dimension (filling with zero the other ones)
+        pad_pads = fill_pads_with_axes(pad_pads, axes, x_rank)
+
+        # Get only spatial pads
+        new_pads = pad_pads[2:x_rank] + pad_pads[x_rank + 2 :]
+
+        # Replace conv pads = new + old
+        conv_attr = conv_node.attributes.copy()
+        if "pads" in conv_attr:
+            new_pads = [x + y for x, y in zip(conv_attr["pads"].as_ints(), new_pads)]
+        conv_attr.add(ir.AttrInt64s("pads", new_pads))
+
+        return op.op(
+            conv_node.op_type,
+            inputs=(x, *conv_node.inputs[1:]),
+            attributes=conv_attr,
+            domain=conv_node.domain,
+            name=conv_node.name,
+        )
+
+    def check(self, context, x: ir.Value, pad: ir.Value, conv: ir.Value) -> orp.MatchResult:
```

Comment:
This is relatively complicated. Could you summarize the check logic here. It would be easier to follow and maintain. Something like this would be helpful: https://github.com/microsoft/onnxscript/blob/b8a831ed49d51da405402bcb10c7b1348197c352/onnxscript/rewriter/broadcast_to_matmul.py#L16-L28